sdk/apimanagement/api-management-custom-widgets-tools/src/node/deploy.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { CONFIG_FILE_NAME } from "../paths";
+import CustomWidgetBlobService from "./CustomWidgetBlobService";
+import fs from "fs";
+import getStorageSasUrl from "./getStorageSasUrl";
+import readdir from "./readdir";
+
+/**
+ * resourceId - resource ID of API Management service "subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.ApiManagement/service/[service-name]"
+ * managementApiEndpoint - URL with protocol (e.g. https://management.azure.com)
+ * apiVersion - optional to override default (e.g. "2019-01-01")
+ * tokenOverride - optional, provides token to use for auth, instead of 'az login' approach
+ */
+export type TServiceInformation = {
+  resourceId: string;
+  managementApiEndpoint: string;
+  apiVersion?: string;
+  tokenOverride?: string;
+};
+
+/**
+ * Deploys everything from /dist folder to the API Management DevPortals' blob storage.
+ *
+ * @param serviceInformation - service information for deployment
+ * @param name - name of the widget to be deployed
+ * @param fallbackConfigPath - local path to the config file (by default "./static/config.msapim.json")
+ */
+async function deploy(
+  serviceInformation: TServiceInformation,
+  name: string,
+  fallbackConfigPath = "./static/" + CONFIG_FILE_NAME
+): Promise<void> {
+  console.log("\n\n");
+  console.log("Starting deploy process of custom widget: " + name);
+  console.log("Please, sign in to your Azure account when prompted\n");
+
+  const blobStorageUrl = await getStorageSasUrl(serviceInformation);
+  const customWidgetBlobService = new CustomWidgetBlobService(blobStorageUrl, name);
+
+  let config;
+  try {
+    console.log("Looking for config file in the Azure blob storage");
+    config = await customWidgetBlobService.getConfig();
+  } catch (e) {
+    console.log("Config not found.");
+  }
+  if (!config) {
+    console.log("Looking for a local config file in: " + fallbackConfigPath);
+    config = JSON.parse(fs.readFileSync(fallbackConfigPath).toString());
+  }
+
+  console.log("Config file loaded\n");
+
+  const rootLocal = "./dist/";
+  const files = readdir("", rootLocal);
+
+  console.log("Starting upload of data files from the '" + rootLocal + "' folder\n");
+
+  await customWidgetBlobService.cleanDataDir();
+
+  const promises: Promise<void>[] = [];
+  files.forEach((file) => {
+    const content = fs.readFileSync(rootLocal + file);
+    const promise = customWidgetBlobService
+      .uploadWidgetDataFile(file, content)
+      .then(() => console.log("Uploaded file: " + file));
```

Comment:
Does this really have an advantage over awaiting the promise to resolve before triggering the new one? In my experience, it is usually the same performance wise and awaiting before kicking off the next one reads better

Reply:
For large project with big files it might make some difference, but I don't really see how is await more readable, it's just slower.
Feel free to refactor it how would you like it more and I'll take a look.

Reply:
Remember that NodeJS is single-threaded, and async doesn't mean parallel. In the end, these tasks need to wait on each other to be processed, so I'm not 100% sure that kicking off all promises and then a promise.all would really improve performance.

And as for readability, it does improve a lot if you await each upload.

```ts
for (const file of files) {
   const content = fs.readFileSync(rootLocal + file);
   await customWidgetBlobService.uploadWidgetDataFile(file, content);
   console.log("Uploaded file: " + file);
}
```